exercises/grep/grep_test.py

```diff
@@ -14,152 +15,105 @@
 To dogs and to all ravening fowls a prey,
 When fierce dispute had separated once
 The noble Chief Achilles from the son
-Of Atreus, Agamemnon, King of men.""",
+Of Atreus, Agamemnon, King of men.\n""",
     "midsummer-night.txt": """I do entreat your grace to pardon me.
 I know not by what power I am made bold,
 Nor how it may concern my modesty,
 In such a presence here to plead my thoughts;
 But I beseech your grace that I may know
 The worst that may befall me in this case,
-If I refuse to wed Demetrius.""",
+If I refuse to wed Demetrius.\n""",
     "paradise-lost.txt": """Of Mans First Disobedience, and the Fruit
 Of that Forbidden Tree, whose mortal tast
 Brought Death into the World, and all our woe,
 With loss of Eden, till one greater Man
 Restore us, and regain the blissful Seat,
 Sing Heav'nly Muse, that on the secret top
 Of Oreb, or of Sinai, didst inspire
-That Shepherd, who first taught the chosen Seed""",
+That Shepherd, who first taught the chosen Seed\n""",
 }
-FILES = {}
 
 
-class File:
-    def __init__(self, name=""):
-        self.name = name
-        self.contents = ""
-
-    def read(self):
-        return self.contents
-
-    def readlines(self):
-        return [line + "\n" for line in self.read().split("\n") if line]
-
-    def write(self, data):
-        self.contents += data
-
-    def __enter__(self):
-        return self
-
-    def __exit__(self, *args):
-        pass
-
-
-# Store builtin definition of open()
-builtins.oldopen = builtins.open
-
-
-def open(name, mode="r", *args, **kwargs):
-    # if name is a mocked file name, lookup corresponding mocked file
-    if name in FILE_TEXT:
-        if mode == "w" or name not in FILES:
-            FILES[name] = File(name)
-        return FILES[name]
-    # else call builtin open()
-    else:
-        return builtins.oldopen(name, mode, *args, **kwargs)
-
-
-# remove mocked file contents
-def remove_file(file_name):
-    del FILES[file_name]
-
-
-def create_file(name, contents):
-    with open(name, "w") as f:
-        f.write(contents)
+def open_mock(fname, *args, **kwargs):
+    try:
+        return io.StringIO(FILE_TEXT[fname])
+    except KeyError:
+        raise RuntimeError(
+            "Expected one of {0!r}: got {1!r}".format(list(FILE_TEXT.keys()), fname)
+        )
 
 
+@mock.patch("grep.open", name="open", side_effect=open_mock)
```

Comment:
Two choices: `@mock.patch("builtins.open", name="open", side_effect=open_mock)` or `@mock.patch("grep.open", name="open", side_effect=open_mock, create=True)`
for it to work everywhere. I forgot that 3.4 doesn't add `create=True` by default for built-in functions the way 3.5 and later do. Note: using the first option will break `open` everywhere for the duration of the test, but that doesn't really matter here as you changed the fallback.

Reply:
Thanks for that, went with the latter. CI now passing.